In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("iris.data", header=None, sep=' ')
df

0    1    2    3  4
0    5.1  3.5  1.4  0.2  1
1    4.9  3.0  1.4  0.2  1
2    4.7  3.2  1.3  0.2  1
3    4.6  3.1  1.5  0.2  1
4    5.0  3.6  1.4  0.2  1
..   ...  ...  ...  ... ..
145  6.7  3.0  5.2  2.3  3
146  6.3  2.5  5.0  1.9  3
147  6.5  3.0  5.2  2.0  3
148  6.2  3.4  5.4  2.3  3
149  5.9  3.0  5.1  1.8  3

[150 rows x 5 columns]

In [3]:
y = df[4] # extraemos la columna de clase
print('Dimensión del vector de clase:',y.shape)
print(y)

Dimensión del vector de clase: (150,)
0      1
1      1
2      1
3      1
4      1
      ..
145    3
146    3
147    3
148    3
149    3
Name: 4, Length: 150, dtype: int64


In [4]:
X= df.drop(4, axis=1) # eliminamos la columna de clase y nos quedamos con el resto
print('Dimensiones de la matriz de datos:',X.shape)
print(X)

Dimensiones de la matriz de datos: (150, 4)
       0    1    2    3
0    5.1  3.5  1.4  0.2
1    4.9  3.0  1.4  0.2
2    4.7  3.2  1.3  0.2
3    4.6  3.1  1.5  0.2
4    5.0  3.6  1.4  0.2
..   ...  ...  ...  ...
145  6.7  3.0  5.2  2.3
146  6.3  2.5  5.0  1.9
147  6.5  3.0  5.2  2.0
148  6.2  3.4  5.4  2.3
149  5.9  3.0  5.1  1.8

[150 rows x 4 columns]


Impresión de la media del DF para observar que devuelve vector de medias. Además, las operaciones de substracción de DF con escalar o valor absoluto, los resultados son un DF.

In [5]:
print(X.mean())

0    5.843333
1    3.054000
2    3.758667
3    1.198667
dtype: float64


In [6]:
print(X - X.mean())

            0      1         2         3
0   -0.743333  0.446 -2.358667 -0.998667
1   -0.943333 -0.054 -2.358667 -0.998667
2   -1.143333  0.146 -2.458667 -0.998667
3   -1.243333  0.046 -2.258667 -0.998667
4   -0.843333  0.546 -2.358667 -0.998667
..        ...    ...       ...       ...
145  0.856667 -0.054  1.441333  1.101333
146  0.456667 -0.554  1.241333  0.701333
147  0.656667 -0.054  1.441333  0.801333
148  0.356667  0.346  1.641333  1.101333
149  0.056667 -0.054  1.341333  0.601333

[150 rows x 4 columns]


In [7]:
print(abs(X - X.mean()))

            0      1         2         3
0    0.743333  0.446  2.358667  0.998667
1    0.943333  0.054  2.358667  0.998667
2    1.143333  0.146  2.458667  0.998667
3    1.243333  0.046  2.258667  0.998667
4    0.843333  0.546  2.358667  0.998667
..        ...    ...       ...       ...
145  0.856667  0.054  1.441333  1.101333
146  0.456667  0.554  1.241333  0.701333
147  0.656667  0.054  1.441333  0.801333
148  0.356667  0.346  1.641333  1.101333
149  0.056667  0.054  1.341333  0.601333

[150 rows x 4 columns]


In [8]:
print(X.std())

0    0.828066
1    0.433594
2    1.764420
3    0.763161
dtype: float64


In [9]:
print(2*X.std())

0    1.656132
1    0.867189
2    3.528841
3    1.526321
dtype: float64


El resultado de la fórmula para detectar 'outliers' es una matriz o DF de booleanos por que
se usa un operador relacional (>=). Esto es los valores que satisfacen la relación son Verdaderos
De lo contrario Falso

In [10]:
n=2
print(abs(X - X.mean()) >= n*X.std())

         0      1      2      3
0    False  False  False  False
1    False  False  False  False
2    False  False  False  False
3    False  False  False  False
4    False  False  False  False
..     ...    ...    ...    ...
145  False  False  False  False
146  False  False  False  False
147  False  False  False  False
148  False  False  False  False
149  False  False  False  False

[150 rows x 4 columns]


Guardamos la matriz de booleanos para detectar o filtrar quiénes son Verdaderos y por lo tanto
son outliers. Después de encontrarlos se procede a eliminarlos

In [11]:
res = abs(X - X.mean()) >= n*X.std()
print(res)
print(res.sum())

         0      1      2      3
0    False  False  False  False
1    False  False  False  False
2    False  False  False  False
3    False  False  False  False
4    False  False  False  False
..     ...    ...    ...    ...
145  False  False  False  False
146  False  False  False  False
147  False  False  False  False
148  False  False  False  False
149  False  False  False  False

[150 rows x 4 columns]
0    6
1    5
2    0
3    0
dtype: int64


La impresión de res.sum() nos indica que la primera y segunda columnas contienen 6 y 5 outliers
respectivamente. Ahora mostramos quiénes son

In [12]:
res[res[0]==True]

0      1      2      3
105  True  False  False  False
117  True  False  False  False
118  True  False  False  False
122  True  False  False  False
131  True  False  False  False
135  True  False  False  False

In [13]:
res[res[1]==True]

0     1      2      3
14  False  True  False  False
15  False  True  False  False
32  False  True  False  False
33  False  True  False  False
60  False  True  False  False

Ya que sabemos quiénes son los outliers, juntamos el total de muestras booleanas en un nuevo DF
para recuperar sus índices de fila con el fin de eliminarlos.

In [14]:
indOut= res[(res[0]==True) | (res[1]==True)]
indOut

0      1      2      3
14   False   True  False  False
15   False   True  False  False
32   False   True  False  False
33   False   True  False  False
60   False   True  False  False
105   True  False  False  False
117   True  False  False  False
118   True  False  False  False
122   True  False  False  False
131   True  False  False  False
135   True  False  False  False

Ahora usamos 'index' para devolver los índices del DF booleano y con 'drop' eliminamos dichos índices de 
filas (los outliers) del DF original X

In [15]:
X= X.drop(indOut.index)
X

0    1    2    3
0    5.1  3.5  1.4  0.2
1    4.9  3.0  1.4  0.2
2    4.7  3.2  1.3  0.2
3    4.6  3.1  1.5  0.2
4    5.0  3.6  1.4  0.2
..   ...  ...  ...  ...
145  6.7  3.0  5.2  2.3
146  6.3  2.5  5.0  1.9
147  6.5  3.0  5.2  2.0
148  6.2  3.4  5.4  2.3
149  5.9  3.0  5.1  1.8

[139 rows x 4 columns]

Para verificar que se realizón con éxito la eliminación de outliers, observe que el nuevo DF X
contiene solamente 139 de las 150 muestras originales, por lo que se eliminaron 11 outliers. Además
los primeros dos outliers estaban en las filas 14 y 15, por lo que al mostrar las primeras 20
muestras se puede ver que ya no se encuentran en el DF X modificado (de la fila 13 se salta a la 16).

In [16]:
X.head(20)

0    1    2    3
0   5.1  3.5  1.4  0.2
1   4.9  3.0  1.4  0.2
2   4.7  3.2  1.3  0.2
3   4.6  3.1  1.5  0.2
4   5.0  3.6  1.4  0.2
5   5.4  3.9  1.7  0.4
6   4.6  3.4  1.4  0.3
7   5.0  3.4  1.5  0.2
8   4.4  2.9  1.4  0.2
9   4.9  3.1  1.5  0.1
10  5.4  3.7  1.5  0.2
11  4.8  3.4  1.6  0.2
12  4.8  3.0  1.4  0.1
13  4.3  3.0  1.1  0.1
16  5.4  3.9  1.3  0.4
17  5.1  3.5  1.4  0.3
18  5.7  3.8  1.7  0.3
19  5.1  3.8  1.5  0.3
20  5.4  3.4  1.7  0.2
21  5.1  3.7  1.5  0.4